### PR TITLE
Add `one` for AbstractString

### PR DESCRIPTION
**Pull request by bdeonovic**
From discussion in https://github.com/JuliaLang/julia/issues/19536

### CODE REVIEW

**Comment by nalimilan** [comment]
Could you add a test for the new method?

**Comment by bdeonovic** [comment]
Does adding a `jldoctest` constitute as adding a test? Or does it have to be added separately somewhere?

**Comment by nalimilan** [comment]
I don't think you need to document the new method at all, since it matches the general docstring about `one`. So just add the test to `test/strings/basic.jl`.

File: base/strings/basic.jl [hunk 1]
```diff
@@ -108,6 +108,8 @@ julia> "Hello " * "world"
 (.*){T<:AbstractString}(v::Vector{T},s::AbstractString) = [i*s for i in v]
 (.*){T<:AbstractString}(s::AbstractString,v::Vector{T}) = [s*i for i in v]
 
+one{U<:AbstractString}(::Type{U}) = convert(U, "")
```
Comment (nalimilan) [inline base/strings/basic.jl]: Sorry for being a pain, but why use `U` when the rest of the file uses `T`?

**Comment by nalimilan** [comment]
Still needs a test.

**Review by nalimilan — APPROVED** [review]
Thanks!

**Comment by bdeonovic** [comment]
I did it! I contributed to open source =D

**Comment by nalimilan** [comment]
Beware, once you start it's hard to stop...

**Comment by tkelman** [comment]
Thanks for the contribution @bdeonovic, but I still think this is a pun and more likely to obscure a bug in code that shouldn't be receiving strings than actually be useful for concatenation. Instead of ptod, concatenating an array of strings would be better as `string(A...)`

**Comment by bdeonovic** [comment]
@tkelman I agree, personally I didn't want to include `*` as concatenation, but since it was there I felt that `one` was needed.

**Comment by oscardssmith** [comment]
A recent issue about using sparse matrices with strings suggests that we should probably have `zero` also.

File: base/strings/basic.jl [hunk 2]
```diff
@@ -108,6 +108,8 @@ julia> "Hello " * "world"
 (.*){T<:AbstractString}(v::Vector{T},s::AbstractString) = [i*s for i in v]
 (.*){T<:AbstractString}(s::AbstractString,v::Vector{T}) = [s*i for i in v]
 
+one{T<:AbstractString}(::Type{T}) = convert(T, "")
```
Comment (stevengj) [inline base/strings/basic.jl]: For consistency, it should probably be:
```
one{T<:AbstractString}(::Union{T,Type{T}}) = convert(T, "")
```
in analogy with e.g. `one(17) == 1`.

**Comment by stevengj** [comment]
+1.  As long as `*` is string concatenation, `one` should be the multiplicative identity.

**Comment by Sacha0** [comment]
> As long as * is string concatenation, one should be the multiplicative identity.

On the one hand I agree with this. On the other I share `@tkelman`'s operator punning sentiments: These issues seem to evidence the anti-punning position in e.g. #11030, rather than providing cause to double down on the pun. Best!

**Review by StefanKarpinski — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
We added the `++` operator so that it could be used for string concatenation, but that never caught on, so we're still using `*` it seems. If someone wants to take on the whole `++` change (or something else entirely), feel free, but until then, this makes sense.

**Comment by ararslan** [comment]
Bump! Will this make it into 0.6?

**Comment by stevengj** [comment]
Yes, I think the consensus was clearly for this as long as `*` is concatenation.

**Pull request merged**